design/service-binding.md

```diff
@@ -0,0 +1,851 @@
+      port: 9092
+```
+
+The Service Binding Operator looks for annotations starting `servicebinding`.
+
+This has the same shortcomings as the first example, with the advantage that the template code is not required in every single `ServiceBindingRequest`.
+
+It would be preferable if Strimzi custom resources were appropriately annotated without the user needing to do it.
+
+## Desired way of consuming binding information
+
+The aim is to make it easy to bind to any service using a `ServiceBindingRequest` CR. Here's an example:
+
+``` yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: my-binding
+spec:
+  services:
+  - group: kafka.strimzi.io
+    kind: Kafka
+    version: v1beta1
+    resourceRef: my-cluster
+```
+
+Then the binding information is supplied in in a standard way, such as environment variables or secrets. For convenience, I have in general chosen to use secrets for binding information (this is controlled by the CSV annotations). Of course, in combination with developer tooling, this can remove the need for the developer to write the code to read credentials and so on. Instead, they're presented to the application as environment variables with well known names.
+
+## Binding data
+
+To connect to Strimzi, a service binding needs the followng:
+
+* **host** - hostname or IP address
+* **port** - port number
+* **userName** - username, optional
+* **password** - password or token, optional
+* **certificate** - certificate, optional
+
+These pieces of information can be provided in a variety of ways:
+
+* annotations on custom resources, indicating `spec` and `status` properties
+* ConfigMap
+* Secret
+
+## Binding to a Kafka cluster with no TLS or authentication
+
+Because Strimzi support multiple listeners and there is also a future plan to enhance the listener configuration capabilities, it seems prudent to design a scheme that works nicely with multiple listeners.
+
+There are two proposals here, with a preference for the first. Consequently, this format is used for the later examples.
+
+### Option 1 - Augment Kafka CR status
+
+This introduces a list of bootstrap addresses into the `status` for a `Kafka` CR like this:
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+ ...
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost1.example.com
+      port: 9092
+    - host: myhost2.example.com
+      port: 9092
+  - type: tls
+    addresses:
+    - host: myhost3.example.com
+      port: 9093
+  bootstrap:
+    plain: myhost1.example.com:9092,myhost2.example.com:9092
+    tls: myhost3.example.com
+```
+
+This has the advantage that the bootstrap servers information is at a known point in the `status` of the `Kafka` custom resource which makes is simple to annotate the CSV so the Service Binding Operator can find it.
+
+Because that structure is a little odd, perhaps this would be preferred.
+
+``` yaml
+bootstrap:
```

Comment:
I really don't understand why the need for new subsection `bootstrap` over the existing `listeners`.

Reply:
This was an intermediate point halfway through debating the syntax of the CSV annotations. I agree.